### PR TITLE
Credentials in firebase

### LINKED ISSUE #52: Still can't leave games sometimes
Screenshot 2020-03-30 at 23 22 16

I don't know how to reproduce it exactly, but this game is stuck. I can't start it and I can't leave it

### LINKED ISSUE #74: Can't perform a React state update on an unmounted component
That's rather low priority, but react complains in console about some asynchronous task being still executed after component unmonuted
Screenshot 2020-04-01 at 09 50 57

### LINKED ISSUE #75: Can't play a game joined on another device
If you join a room on one device and then switch to another one you can't continue to play this game.
You can join the room but none of your moves get processed. We must either fix it or disallow joining game from different device

### PR DESCRIPTION
**Pull request by pastDexter**
I've changed quite a lot of things. 🙈  Here is a quick summary:
- Removed `UserClient` and moved it to stateless `Auth` module
- Moved all Firestore operations to stateless `DataStore` module
- State management of `user` and authentication state has been moved to `UserContext`
- No more direct usage of `UserContext` - extracted custom `use*` hooks
- Change `users` firebase collection to `public_profiles` to indicate that it's publicly available to everyone.
- Rename `nickname` to `displayName` and `avatar` to `photoURL` to keep naming consistent with firebase user properties. Store this data in private auth user profile (along with email etc), and copy those two publicly available fields to `public_profiles`.
- Generate `photoURL` (avatar) on user creation
- Fetch whole `public_profiles` collection and subscribe to all changes in `UserContext` to keep list of profiles up to date and avoid fetching individual nicknames every time it's needed.
- Create `game_credentials` firestore collection to store private access credentials to each game room. Subscribe to changes in this collection to keep user's credentials map up to date.
- Refactor `GamePage` joining logic to be more "synchronous" and prevent multiple joining and checking request to happen at once.

Closes #52
Closes #74
Closes #75

### CODE REVIEW

File: src/games/kalambury/sidebar/Player.js [hunk 1]
```diff
@@ -1,25 +1,24 @@
...
   handleGuessClick,
 }) => {
   const { t } = useTranslation("kalambury");
-  const { getNickName } = useContext(UserContext);
-  const { data: nickname } = useQuery(["user-nickname", name], (key, id) => getNickName(id));
-  if (!name || empty) {
+  const profiles = useProfiles();
+  const { displayName, photoURL } = profiles.get(uid);
+
+  if (!uid || empty) {
```
Comment (Bombasarkadian) [inline src/games/kalambury/sidebar/Player.js:21]: Do we use `empty` anymore?

File: src/games/kalambury/Sidebar.js [hunk 2]
```diff
@@ -1,54 +1,54 @@
...
+              <Label.Detail>{t("game.public")}</Label.Detail>
+            </Label>
+          )}
+        </Header.Subheader>
+      </Header>
+      <Segment.Group style={{ marginTop: "-5px" }}>
+        {gameMetadata.map(({ id, name }) => (
+          <Player
+            key={id}
+            uid={name}
```
Comment (Bombasarkadian) [inline src/games/kalambury/Sidebar.js:38]: Is it `uid` or `name`? 👀

File: src/games/kalambury/sidebar/Player.js [hunk 3]
```diff
@@ -1,25 +1,24 @@
...
   isWinning,
   isCurrentPlayer,
   handleGuessClick,
 }) => {
   const { t } = useTranslation("kalambury");
-  const { getNickName } = useContext(UserContext);
-  const { data: nickname } = useQuery(["user-nickname", name], (key, id) => getNickName(id));
-  if (!name || empty) {
+  const profiles = useProfiles();
+  const { displayName, photoURL } = profiles.get(uid);
```
Comment (Bombasarkadian) [inline src/games/kalambury/sidebar/Player.js:19]: Maybe we could pass those 2 as props, so this component doesn't have to know about how we persist user data?

File: src/games/kalambury/Sidebar.js [hunk 4]
```diff
@@ -1,54 +1,54 @@
...
-          isCurrentPlayer={pid === playerID}
-          handleGuessClick={handleGuessClick}
-          {...playersData[pid]}
-        />
-      ))}
-      {Array(numPlayers - Object.keys(playersData).length)
-        .fill(0)
-        .map((_, idx) => (
-          <Player key={"dummy" + idx} empty={true} />
+  gameMetadata,
```
Comment (Bombasarkadian) [inline src/games/kalambury/Sidebar.js:12]: Storybooks are now broken 😞

File: src/games/kalambury/Sidebar.js (hunk 2 above)
Comment (pastDexter) [reply src/games/kalambury/Sidebar.js:38]: it's `name` in Lobby API and `uid` in Firebase 😫

File: src/games/kalambury/sidebar/Player.js (hunk 3 above)
Comment (pastDexter) [reply src/games/kalambury/sidebar/Player.js:19]: 👍

File: src/games/kalambury/sidebar/Player.js (hunk 1 above)
Comment (pastDexter) [reply src/games/kalambury/sidebar/Player.js:21]: no, we don't 👍

File: src/games/kalambury/Sidebar.js (hunk 2 above)
Comment (Bombasarkadian) [reply src/games/kalambury/Sidebar.js:38]: Maybe we could map one of them somewhere in-between (somewhere different that when passing to a component 😃)?

Comment (Bombasarkadian) [reply src/games/kalambury/Sidebar.js:38]: But there doesn't seem to be any obvious place of doing it 😕

**Pull request merged**